### PR TITLE
Support MqttPublishMessage

### PR DESCRIPTION
**Pull request by Lico-Tom**
publish message to pulsar Lico-Tom

### CODE REVIEW

File: src/test/java/com/github/shoothzj/mjp/client/Mqttv3ClientTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+@Slf4j
+public class Mqttv3ClientTest {
```
Comment (goflutterjava) [inline src/test/java/com/github/shoothzj/mjp/client/Mqttv3ClientTest.java:34]: we already merge the v3Client test

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 2]
```diff
@@ -176,7 +178,69 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+            log.error("FAILURE");
+            return;
+        }
+
+        int len = msg.payload().readableBytes();
+        byte[] messageBytes = new byte[len];
+
+        msg.payload().getBytes(msg.payload().readerIndex(), messageBytes);
+        if (log.isDebugEnabled()) {
+            log.debug("publish message : {} ", new String(messageBytes));
```
Comment (goflutterjava) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:198]: in any case, we will not print the message content

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 3]
```diff
@@ -176,7 +178,69 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+        }
+
+        MqttSessionKey mqttSessionKey = ChannelUtils.getMqttSession(ctx.channel());
```
Comment (goflutterjava) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:201]: mqttSessionKey may be null

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 4]
```diff
@@ -176,7 +178,69 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+                }
+            } catch (PulsarClientException e) {
+                log.error("fail to send message to pulsar : {} ", e.getMessage());
+            }
+        }
+    }
+
+    private void processProducer(MqttSessionKey mqttSessionKey, String topic) throws PulsarClientException {
+        List<MqttTopicKey> mqttTopicKeys = sessionProducerMap.get(mqttSessionKey);
+        if (mqttTopicKeys.size() != 0) {
```
Comment (goflutterjava) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:224]: this part should protected by lock

**Review by goflutterjava — CHANGES_REQUESTED** [review]

**Comment by goflutterjava** [comment]
@fu-turer @zxJin-x PTAL, thx

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 5]
```diff
@@ -176,7 +178,69 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+            }
+            return;
+        }
+        MqttTopicKey mqttTopicKey = new MqttTopicKey();
+        mqttTopicKey.setMqttSessionKey(mqttSessionKey);
+        mqttTopicKey.setTopic(topic);
+        ArrayList<MqttTopicKey> topics = Lists.newArrayList();
+        topics.add(mqttTopicKey);
+        sessionProducerMap.put(mqttSessionKey, topics);
+        producerMap.put(mqttTopicKey, pulsarClient.newProducer()
```
Comment (fu-turer) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:241]: close producer in `processDisconnect`

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 6]
```diff
@@ -176,7 +178,69 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+        }
+
+        MqttSessionKey mqttSessionKey = ChannelUtils.getMqttSession(ctx.channel());
+        String topic = mqsarServer.produceTopic(mqttSessionKey.getUsername(),
+                mqttSessionKey.getClientId(), msg.variableHeader().topicName());
+        processProducer(mqttSessionKey, topic);
+        for (MqttTopicKey mqttTopicKey : sessionProducerMap.get(mqttSessionKey)) {
+            try {
+                if (topic.equals(mqttTopicKey.getTopic())) {
+                    producerMap.get(mqttTopicKey).send(messageBytes);
```
[the remark this answers is not in the document]
Comment (Shoothzj) [reply src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:208]: Maybe we should judge the qos first, and introduce a config like Qos1NoWait to control if qos1 sendAsync.

Qos0 should sendAsync. I agree

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 7]
```diff
@@ -176,7 +178,67 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+            return;
+        }
+
+        int len = msg.payload().readableBytes();
+        byte[] messageBytes = new byte[len];
+
+        msg.payload().getBytes(msg.payload().readerIndex(), messageBytes);
+        MqttSessionKey mqttSessionKey = ChannelUtils.getMqttSession(ctx.channel());
+        if (mqttSessionKey == null) {
+            mqttSessionKey = new MqttSessionKey();
```
Comment (goflutterjava) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:199]: we should fail and disconnect if mqttSession key be null

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 8]
```diff
@@ -176,7 +178,67 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+        }
+        String topic = mqsarServer.produceTopic(mqttSessionKey.getUsername(),
+                mqttSessionKey.getClientId(), msg.variableHeader().topicName());
+        Producer producer = getOrCreateProducer(mqttSessionKey, topic);
+        if (msg.fixedHeader().qosLevel() == MqttQoS.AT_MOST_ONCE) {
+            producer.sendAsync(messageBytes);
```
[the remark this answers is not in the document]
Comment (goflutterjava) [reply src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:205]: yes, we should handle the callback to print log to locate problems

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 9]
```diff
@@ -176,7 +178,67 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+        }
+        if (msg.fixedHeader().qosLevel() == MqttQoS.AT_LEAST_ONCE) {
+            producer.send(messageBytes);
+            MqttPubAckMessage pubAckMessage = (MqttPubAckMessage) MqttMessageFactory.newMessage(
```
Comment (goflutterjava) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:209]: we should handle the send result, if not success, we will not send the `pubAck` msg

File: src/test/java/com/github/shoothzj/mjp/client/Mqttv3ClientTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.github.shoothzj.mjp.client;
+
+import com.github.shoothzj.mjp.MqsarBroker;
+import com.github.shoothzj.mjp.integrate.MqsarTestUtil;
```
Comment (goflutterjava) [inline src/test/java/com/github/shoothzj/mjp/client/Mqttv3ClientTest.java:23]: we do not need commit this file, commits need squash

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 11]
```diff
@@ -44,15 +45,14 @@
 import io.netty.handler.codec.mqtt.MqttConnAckMessage;
 import io.netty.handler.codec.mqtt.MqttUnacceptableProtocolVersionException;
 import io.netty.handler.codec.mqtt.MqttMessageFactory;
+import io.netty.handler.codec.mqtt.MqttMessageIdVariableHeader;
 import lombok.extern.slf4j.Slf4j;
 import org.apache.commons.compress.utils.Lists;
-import org.apache.pulsar.client.api.Consumer;
-import org.apache.pulsar.client.api.Producer;
+import org.apache.pulsar.client.api.*;
```
Comment (Shoothzj) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:51]: avoid use start import

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 12]
```diff
@@ -176,7 +176,80 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+            log.error("FAILURE");
+            return;
+        }
+
+        int len = msg.payload().readableBytes();
+        byte[] messageBytes = new byte[len];
+
+        msg.payload().getBytes(msg.payload().readerIndex(), messageBytes);
+        MqttSessionKey mqttSessionKey = ChannelUtils.getMqttSession(ctx.channel());
+        Preconditions.checkNotNull(mqttSessionKey);
```
Comment (Shoothzj) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:196]: don't assert it be null, if null, we need to close the channel

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 13]
```diff
@@ -176,7 +176,80 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
     void processPubAck(ChannelHandlerContext ctx, MqttPubAckMessage msg) {
     }
 
-    void processPublish(ChannelHandlerContext ctx, MqttPublishMessage msg) {
+    void processPublish(ChannelHandlerContext ctx, MqttPublishMessage msg) throws PulsarClientException {
```
Comment (Shoothzj) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:179]: what's the behavior if we throws `PulsarClientException ` here

File: src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java [hunk 14]
```diff
@@ -176,7 +176,80 @@ void processConnect(ChannelHandlerContext ctx, MqttConnectMessage msg) {
...
+
+        if (msg.fixedHeader().qosLevel() == MqttQoS.EXACTLY_ONCE) {
+            log.error("Does not support to QoS2 protocol");
```
Comment (Shoothzj) [inline src/main/java/com/github/shoothzj/mjp/MqsarProcessor.java:182]: should print the mqtt clientId

**Pull request closed without merging**